ArduCopter/events.cpp

```diff
@@ -4,152 +4,221 @@
  *       This event will be called when the failsafe changes
  *       boolean failsafe reflects the current state
  */
+
+bool Copter::failsafe_option(FailsafeOption opt) const
+{
+    return (g2.fs_options & (uint32_t)opt);
+}
+// enum Failsafe_Action {
+//     Failsafe_Action_None           = 0,
+//     Failsafe_Action_Land           = 1,
+//     Failsafe_Action_RTL            = 2,
+//     Failsafe_Action_SmartRTL       = 3,
+//     Failsafe_Action_SmartRTL_Land  = 4,
+//     Failsafe_Action_Terminate      = 5
+// };
 void Copter::failsafe_radio_on_event()
 {
-    // if motors are not armed there is nothing to do
-    if( !motors->armed() ) {
-        return;
-    }
+    AP::logger().Write_Error(LogErrorSubsystem::FAILSAFE_RADIO, LogErrorCode::FAILSAFE_OCCURRED);
 
+    int16_t desired_action = g.failsafe_throttle; //The action set in FS_THR_ENABLE
+
+    // Conditions to deviate from FS_THR_ENABLE selection and sent specific GCS warning
     if (should_disarm_on_failsafe()) {
+        // should immediately disarm when we're on the ground
+        gcs().send_text(MAV_SEVERITY_WARNING, "Radio Failsafe - Disarming");
         arming.disarm();
+        desired_action = FS_THR_DISABLED;
+
+    } else if (flightmode->is_landing() && ((battery.has_failsafed() && battery.get_highest_failsafe_priority() <= FAILSAFE_LAND_PRIORITY))) {
+        // Allow landing to continue when battery failsafe requires it (not a user option)
+        gcs().send_text(MAV_SEVERITY_WARNING, "Radio + Battery Failsafe - Continuing Landing");
+        desired_action = FS_THR_ENABLED_ALWAYS_LAND;
+
+    } else if (flightmode->is_landing() && failsafe_option(FailsafeOption::CONTINUE_IF_LANDING)) {
+        // Allow landing to continue when FS_OPTIONS is set to continue landing
+        gcs().send_text(MAV_SEVERITY_WARNING, "Radio Failsafe - Continuing Landing");
+        desired_action = FS_THR_ENABLED_ALWAYS_LAND;
+
+    } else if (control_mode == Mode::Number::AUTO && failsafe_option(FailsafeOption::RC_CONTINUE_IF_AUTO)) {
+        // Allow mission to continue when FS_OPTIONS is set to continue mission
+        gcs().send_text(MAV_SEVERITY_WARNING, "Radio Failsafe - Continuing Auto Mode");       
+        desired_action = FS_THR_DISABLED;
+
+    } else if ((flightmode->in_guided_mode()) &&
+      (failsafe_option(FailsafeOption::RC_CONTINUE_IF_GUIDED)) && (g.failsafe_gcs != FS_GCS_DISABLED)) {
+        // Allow guided mode to continue when FS_THR_OPTION is set to continue in guided mode.  Only if the GCS failsafe is enabled.
```

Comment:
I guess it's FS_OPTIONS instead of FS_THR_OPTION?

Reply:
Fixed.